Drop Python 2 support

jdemel opened the pull request:
The minimum required Python version is 3.4 now. Before it was either Python2.7 or Python3.4+. It is 2020 and the pythonclock ran out, thus we drop Python2 support.   

With Python2 support dropped, we can also
- remove `six` as a dependency.
- replace `optparse` by `argparse` because `optparse` is deprecated.
- remove `from __future__` imports because they are obsolete now.

CI update
- Remove obsolete packages, `six`, `Cheetah`, `python2`. 
- Move to newer distro 'bionic' by default.

Update: I broke the CI system. Thus, I removed all CI system changes. This would probably be better as a separate PR.

jdemel committed:
python: Drop Python2 support

The minimum required Python version is 3.4 now. Before it was either
Python2.7 or Python3.4+. It is 2020 and the pythonclock ran out, thus we
drop Python2 support.

jdemel committed:
python: Reduce six usage

With Python2 support dropped, we can also remove 'six' as a dependency.

jdemel committed:
python: Move to Python3 syntax and modules

This commit removes six usage. Also, optparse is mostly replaced by
argparse because optparse is deprecated. All `from __future__` imports
are removed because they are obsolete now.

jdemel commented:
AppVeyor fails because it is missing Python3 as far as I can tell. If someone can suggest how to fix this, that would be awesome.

noc0lour commented:
For appveyor we need to check how to use this: https://www.appveyor.com/docs/windows-images-software/#python to use python3.x
and for the buildbot I don't really know if we need it for volk really, as travis-ci is already there.

noc0lour commented:
Maybe there are mutual exclusive SIMD instruction sets, which would increase test coverage, but other than that, I don't know.

noc0lour commented:
Anyhow an update to Python3 is needed in the CI for volk if we want to continue use the GR CI.

michaelld reviewed (approved):
LGTM ;  before merging, let's get the CI working again to verify more robustly

velichkov committed:
appveyor: Test with python 3.4 and 3.8

velichkov committed:
appveyor: Add job names

velichkov commented:
@jdemel For appveyor the following change (0a7349e) should enable python3

```diff
@@ -10,9 +10,11 @@ environment:
     - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2017
       CMAKE_GENERATOR: Visual Studio 12 2013
       VC_VERSION: vc120
+      PYTHON: "C:\\Python34-x64"
     - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2017
       CMAKE_GENERATOR: Visual Studio 14 2015
       VC_VERSION: vc140
+      PYTHON: "C:\\Python38-x64"
 
 install:
     - nuget install boost_system-%VC_VERSION% -Version %BOOST_VERSION% -OutputDirectory %CD%/packages
@@ -20,7 +22,13 @@ install:
     - nuget install boost_chrono-%VC_VERSION% -Version %BOOST_VERSION% -OutputDirectory %CD%/packages
     - nuget install boost_program_options-%VC_VERSION% -Version %BOOST_VERSION% -OutputDirectory %CD%/packages
     - nuget install boost_unit_test_framework-%VC_VERSION% -Version %BOOST_VERSION% -OutputDirectory %CD%/packages
-    - pip install Cheetah
+    # Prepend the selected Python to the PATH of this build
+    - SET PATH=%PYTHON%;%PYTHON%\Scripts;%PATH%
+    # Display version information about selected python and pip
+    - python --version
+    - python -c "import sys, platform, struct;
+      print(sys.platform, platform.machine(), struct.calcsize('P')*8)"
+    - pip --version
     - pip install mako
     - pip install six
 before_build:
```
The cmake still checks for `six` so CMakeLists.txt still needs to be fixed. I've successfully tested them in [my appveyor account](https://ci.appveyor.com/project/velichkov/volk/builds/30387109). In 32533b0 I've added short job names.

Let me know if you are going to incorporate these changes in this PR or you prefer a separate PR against master.

jdemel commented:
@velichkov Since we rely on CI I'd like to incorporate your change in this PR. I just picked your 2 commits that update `appveyor.yml` and pushed them. I hope this will be a success and we are one step further with this PR.
Thanks for your work!

jdemel committed:
appveyor: Make ctest more verbose

ctest is called with `-V` now to make it more verbose. In case of errors
this might be helpful.

jdemel committed:
six: Remove build dependency on python six

Since we require Python3 for builds now, we can remove python six
as a dependency. We no longer need cross-compatibility between Py2k
and Py3k.

michaelld commented:
@jdemel let me know when you're satisfied & I'll finalize my review. everything still LGTM ...

jdemel commented:
@michaelld I think this PR is finished. I'm currently happy with it as it is.

michaelld reviewed (approved):
LGTM

michaelld commented:
If you missed anything Py2 related, we can continue cleaning it up in another PR. This one is pretty comprehensive IMHO.

michaelld commented:
CI that fails has nothing to do with this PR. CI does pass as desired. Merging.
